pymatgen/io/res.py

```diff
@@ -0,0 +1,660 @@
+"""
+Provides parsing and read/write support for ShelX .res files as produced by the AIRSS code.
+
+Converting from and back to pymatgen objects is expected to be reversible, i.e. you
+should get the same Structure or ComputedStructureEntry back. On the other hand, converting
+from and back to a string/file is not garunteed to be reversible, i.e. a diff on the output
+would not be empty. The difference should be limited to whitespace, float precision, and the
+REM entries.
+
+"""
+
+import datetime
+import re
+from dataclasses import dataclass
+from typing import Callable, Dict, Iterator, List, Literal, Optional, Set, Tuple, Union
+
+import dateutil.parser  # type: ignore
+from monty.io import zopen
+
+from pymatgen.core.lattice import Lattice
+from pymatgen.core.periodic_table import Element
+from pymatgen.core.sites import PeriodicSite
+from pymatgen.core.structure import Structure
+from pymatgen.entries.computed_entries import ComputedStructureEntry
+
+__all__ = ["ResProvider", "AirssProvider", "ResIO", "ResWriter", "ParseError", "ResError"]
+
+
+@dataclass(frozen=True)
+class AirssTITL:
+    seed: str
+    pressure: float
+    volume: float
+    energy: float
+    integrated_spin_density: float
+    integrated_absolute_spin_density: float
+    spacegroup_label: str
+    appearances: int
+
+    def __str__(self) -> str:
+        title_fmt = "TITL {:s} {:.2f} {:.4f} {:.5f} {:f} {:f} ({:s}) n - {:d}"
+        return title_fmt.format(
+            self.seed,
+            self.pressure,
+            self.volume,
+            self.energy,
+            self.integrated_spin_density,
+            self.integrated_absolute_spin_density,
+            self.spacegroup_label,
+            self.appearances,
+        )
+
+
+@dataclass(frozen=True)
+class ResCELL:
+    unknown_field_1: float
+    a: float
+    b: float
+    c: float
+    alpha: float
+    beta: float
+    gamma: float
+
+    def __str__(self) -> str:
+        cell_fmt = "CELL {:.5f} {:.5f} {:.5f} {:.5f} {:.5f} {:.5f} {:.5f}"
+        return cell_fmt.format(self.unknown_field_1, self.a, self.b, self.c, self.alpha, self.beta, self.gamma)
+
+
+@dataclass(frozen=True)
+class Ion:
+    specie: str
+    specie_num: int
+    pos: Tuple[float, float, float]
```

Comment:
The rest of pymatgen uses

```py
from __future__ import annotations
```

if you add that at the very top and run `pyugrade` on this file, you can auto-convert all type hints to e.g.

```py
pos: tuple[float, float, float]
species: set[str]
ions: list[Ion]
```